### ISSUE TITLE
Make it more obvious when the sx prop is used

### ISSUE THREAD

**Issue by oliviertassinari** [mui/mui-toolpad#787]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Summary 💡

I'm reporting a small frustration I had. If you open https://master--toolpad.mui.com/_toolpad/app/cl42qt2jx01249xnr14xjce8b/editor/pages/cl6r0xb1f00003b6ghdh8e37l can you figure out why the text field is not full width? 😁 It's coming from 

```json
{
  "width": 175
}
```

**Comment by Janpot** [comment]
Also related to https://github.com/mui/mui-toolpad/issues/38

**Comment by prakhargupta1** [comment]
User feedback: if I didn't know "sx" was a MUI thing, I'd see it on the sidebar and think "size?"
So we should link the sx prop page here to help developers understand the concept better.

**Comment by bytasv** [comment]
Maybe also rename it in UI and call it `style` or something more human readable?

**Comment by Janpot** [comment]
> So we should link the sx prop page here to help developers understand the concept better.

We are linking the sx prop page:

<img width="630" alt="Screenshot 2023-01-11 at 10 26 50" src="https://user-images.githubusercontent.com/2109932/211768584-754bfefe-e878-4ba3-87c1-0cb564a884e1.png">

There is a [tracking issue](https://github.com/mui/mui-toolpad/issues/38) to create a more human friendly version of it.

**Comment by prakhargupta1** [comment]
Q1. Somehow the above helper text, remains longer than intended. Having it accessible through a `?` icon would solve this problem


https://user-images.githubusercontent.com/92228082/214042574-db7b230c-aced-4174-91bb-bf283ecb5c18.mov

Q2. Like the way we show 2 props already setup for the container component, is it possible to show it for all components? We can show the default styling values. It will help the user quickly play around some of them and understand how styling works here.

**Comment by bytasv** [comment]
> Q2. Like the way we show 2 props already setup for the container component, is it possible to show it for all components?

container component has these defaults defined by us, so it's not a container styling props, rather it's overrides. We can show these props for other components too, but first they have to be defined individually

**Comment by prakhargupta1** [comment]
> container component has these defaults defined by us, so it's not a container styling props, rather it's overrides. We can show these props for other components too, but first they have to be defined individually

Understood. I'll try and define these material design defaults can create as issue for it later.

**Comment by gerdadesign** [comment]
It seems like there's a couple factors here. For the scope of this issue, below is a suggestion for how to visually display when the sx prop is being used. 

- Use a toggle to show when the sx prop is on/off
- The databinding link icon can still be used as an independent indicator
- I also felt the language in the tooltip is a lot to read in the current state, so attempted a cut version. 
  - I opted to keep it in the tooltip vs a separate help icon because this is consistent with the pattern as every other item on the panel.
  - This still includes the link to the SX Prop docs. 
  - I renamed the toggle to give more information, but if users have an issue with discoverability between that and the tooltip, I am open to reworking in the future to address that. 
- I believe that the toggle is a flexible option: 
  - For now, enabling the toggle can open the existing dialog 
  - In the future (#38), we have options to then either adjust the dialog content or make additional options visible underneath the toggle. 

### Default 
![default](https://user-images.githubusercontent.com/106991024/215149663-437f3ba1-909c-494d-a536-50bf7bb57042.jpg)

### Hovering (tooltip)
![hover](https://user-images.githubusercontent.com/106991024/215149633-c3908a32-3623-4dd0-8124-b4ff98c7cf45.jpg)

### SX prop is being used 
![enabled](https://user-images.githubusercontent.com/106991024/215149659-6668ba67-060c-42f3-8877-45367e62d773.jpg)

### Adding data binding automatically means that SX prop is being used 
![databinding](https://user-images.githubusercontent.com/106991024/215149655-cff1b3b1-747a-4c9a-aa75-fadb391494ba.jpg)

**Comment by Janpot** [comment]
A switch without immediate feedback feels a bit against the [MD guidelines](https://m3.material.io/components/switch/guidelines). (Not that we have to be religious about this as far as I'm concerned).

Could there be alternatives similar to how we did the binding button? inactive/active state + different icon?

**Comment by oliviertassinari** [comment]
> t's a bit counterintuitive to me when clicking a switch opens a dialog.

Agree, a toggle can't open a dialog. e.g. to edit you then need to disable -> reenable?

I think that we miss a visual indicator, e.g.:

<img width="372" alt="Screenshot 2023-01-29 at 18 59 04" src="https://user-images.githubusercontent.com/3165635/215346119-8d9f4786-ef2d-4998-81ae-764c1fdd0636.png">

where the label is "Edit" or "Add" depending on the state.

**Comment by Janpot** [comment]
Also want to add, this doesn't necessarily _have_ to be a single line/button. e.g We could think up a larger component containing 5 nested editors for the 5 most used css properties, and a button to open an editor in a modal/popover for all the other properties.

**Comment by prakhargupta1** [comment]
I agree with Olivier's point that separating title and action will remove the confusion about whether the sx prop is being used or not. About how to show the action Add/Edit, I was thinking if we should show an edit (pencil) icon. Benefits:

1. It will be consistent with the binding icon.
2. States like used/error/not used can be shown like we show for binding icon.
3. The same pattern can be extended to other component props like a) Datagrid: rows, selection  b) Select: options

Also, we need to ensure that a pattern is followed for all possible combinations in the right sidebar. We have text input, drop-down, buttons etc. We have a separate task for this: https://github.com/mui/mui-toolpad/issues/1488

**Comment by apedroferreira** [comment]
+ 1 for feeling like a switch is not the best solution, it feels more appropriate for something that can only be true/false
Also yes, it might be a good idea if every control in the right panel that opens a modal follows the same type of pattern.

**Comment by Janpot** [comment]
> On the topic of the tooltip helpers:

I think a hybrid of those could make sense. e.g. either
- Short description + "click for more info" in popover, open on hover, and opening the docs on click. 
- Short description in popover, open on click, "more info" link inside the popover to open the docs

I also like that both of these examples are quite discoverable with a sort of ⓘ icon. I feel like that's missing the most from Toolpad at the moment. The current components aren't very dense and the label doesn't really allow for adding an icon like that.

**Comment by gerdadesign** [comment]
Question about this SX prop within the context of the new direction. If our target user is a backend dev, it does not seem that they would know how to use the SX prop, nor want to read through the documentation to understand how to use it. Also, would it be possible to bind data to each individual property within the SX prop or only for the SX prop itself? 
I'm curious what makes most sense from the options I see here: 

**1.  Button** 
Similar to the existing experience — when nothing is bound, selecting "add" would open either A. a dialog OR B. bring you to a specific section in your code editor (if going, with this option 1, does A or B make more sense?) If you already have something bound, it would changed Add → Edit.  
<img src="https://user-images.githubusercontent.com/106991024/225019247-d1a1b465-df91-4714-8a17-4b4daf6569f6.jpg" alt="Toolpad_SX_button" width="300">

**2. Editor**
Instead of a pop-up dialog, the editor is directly accessible without leaving the panel. This would make it faster to access, but could get crowded if there is a lot of code.  
<img src="https://user-images.githubusercontent.com/106991024/225019259-b58f40ff-ddc8-4727-a599-af76539546c6.jpg" alt="Toolpad_SX_editor" width="300">

**3. Exposed props**
Assuming most backend devs don't know about the SX prop nor want to learn it, should the most common/relevant properties be more exposed? plus an option to edit via code? This would allow for greater clarity of what you can adjust, but could feel limited (although we would always provide an escape hatch). This primarily necessitates the question of is it possible to bind each property (margin, padding, color, etc) or do you bind the entire sx prop overall? 
<img src="https://user-images.githubusercontent.com/106991024/225019261-07da57f9-42a7-41ea-a546-beca2f6e5d87.jpg" alt="Toolpad_SX_exposedprops" width="300">

Or, if there is a different way of showing this that makes more sense, eager to hear your ideas.

**Comment by Janpot** [comment]
Personally, 3. is what I would advocate for going forward. Some notes:
- It could help in preventing certain properties to be updated by the user if they would mess too much with the Toolpad runtime. e.g. We wouldn't want anyone to make someting `position: 'absolute'`
- UI-wise, we should be mindful that the amount of properties that can be changed, which could be a lot over time.
- I'm not sure whether we should make them bindable. At least not all of them. e.g. I feel like allowing people to bind padding or margin could be a footgun, unlike color.
- There's another option that could be more free form than individual property controls and more restricted than a JSON editor. Similar to how how the browser devtools allow CSS updates:

    https://user-images.githubusercontent.com/2109932/225073527-20b196a4-75ec-46c9-9537-d97bf752a03b.mov

    Or similar to how we allow setting headers/parameters in queries

**Comment by gerdadesign** [comment]
Yeah, I was kind of wishing for something like this in the current iteration, at the very least to help with formatting. I may know I'm looking for `border-color`, or `text-decoration` but used to the CSS formatting, so finding it in JSON takes extra time (just because I don't know). 

If we go with something more freeform like this, could we provide an autocomplete (in JSON format?) with all available properties? Or only properties for which we want to support binding? Can we specify which we do/do not allow?

**Comment by prakhargupta1** [comment]
I think 3. makes sense as this will mean lesser code writing in the UI editor resulting in better UX than having autocomplete in an editor. Choosing from a list of drop-down options, typing input fields would make app building faster.

Also, for a more detailed component, the user anyway would have the option to create a custom component that can have inline sx/styling. So either pro-code or low-code.

But the dev tools vide looks good too, all props are getting validated inline. If maintaining this can be faster, something like this in our JSON editor might be a good way to start and keep UX clean, minimal.
